# 4.5 - Data Consistency Checks

### Table of Contents

* [Importing libraries](#import_libraries)
* [Importing datasets into pandas dataframe](#import_data)
* [Data Consistency Checks](#data_checks)
    * [Checking for Missing Values](#missing_values)
    * [Checking for Duplicates](#duplicates)
*[Tasks for Exercise 4.5](#task_4_5)
    * [Step 2](#step2)
    * [Step 3](#step3)
    * [Step 5](#step5)
    * [Step 6](#step6)
    * [Step 7](#step7)
    * [Step 9](#step9)

# Importing libraries <a class="anchor" id="import_libraries"></a>

In [1]:
#Import pandas, numpy and os libraries
import pandas as pd
import numpy as np
import os

# Importing datasets into pandas dataframe <a class="anchor" id="import_data"></a>

In [2]:
#Creating a shortcut for folderpath for importing data sets
path = r'/Users/kerimzengin/Desktop/Data Analytics/Data Analytics Immersion/Achievement 4 - Python Fundamentals for Data Analysts/Instacart Basket Analysis/Data'

In [3]:
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)

# Data Consistency Checks <a class="anchor" id="data_checks"></a>

## Checking for Missing Values <a class="anchor" id="missing_values"></a>

In [5]:
#checking missing values counts in the Product dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
#Creating dataframe containing all missing values within the product_name column
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [7]:
#showing results for all missing values
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [8]:
#Checking the number of rows before removing rows with missing values for Products dateframe
df_prods.shape

(49693, 5)

In [9]:
#Creating a new dataframe for the cleaned Products dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [10]:
#Checking the number of rows after removing rows with missing values for Products dataframe
df_prods_clean.shape

(49677, 5)

## Checking for Duplicates <a class="anchor" id="duplicates"></a>

In [11]:
#Checking the full duplicates in the cleaned version of Products dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [12]:
#Showing the duplicated records in the Products dataframe
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [13]:
#Creating a new dataframe to not include the duplicated records
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [14]:
#Checking the number of rows after removing duplicates
df_prods_clean_no_dups.shape

(49672, 5)

# Tasks for Exercise 4.5 <a class="anchor" id="task_4_5"></a>

# 2. Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
## Tip: Keep an eye on min and max values! <a class="anchor" id="step2"></a>

In [15]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### The max value for prices is very high at 99999.

# 3. Check for mixed-type data in your df_ords dataframe. <a class="anchor" id="step3"></a>

In [16]:
#Checking for mixed data type for the Orders dataframe
for col in df_ords.columns.tolist():
  mixed_types = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[mixed_types]) > 0:
    print (col)

#### No mixed type columns were found

# 5. Run a check for missing values in your df_ords dataframe.
## In a markdown cell, report your findings and propose an explanation for any missing values you find. <a class="anchor" id="step5"></a>

In [17]:
#Checking for missing values within the Order dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_days_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### The column days_since_prior_order has a very siginificantly high number of missing values, possibly this could be that a lot of customers only made 1 order in total.

# 6. Address the missing values using an appropriate method.
## In a markdown cell, explain why you used your method of choice. <a class="anchor" id="step6"></a>

In [18]:
#Checking the number of rows before removing missing values for Orders dataframe
df_ords.shape

(3421083, 8)

In [19]:
#Creating a new dataframe with non missing values, removing missing values within Orders dataframe
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

#### I have decided to create a new dataframe instead of overwriting the current dataframe, since the number of missing values with really high, it would have a big impact on any future analysis.

In [20]:
#Checking the number of rows after removing missing values for Orders dataframe
df_ords_clean.shape

(3214874, 8)

# 7. Run a check for duplicate values in your df_ords data.
## In a markdown cell, report your findings and propose an explanation for any duplicate values you find. <a class="anchor" id="step7"></a>

In [21]:
#Checking for full duplicates for the cleaned Orders dataframe
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [22]:
#Showing any duplicated records after duplicate checks.
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_days_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### There are no duplicate records found.

# 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names. <a class="anchor" id="step9"></a>

In [23]:
df_ords_clean.to_csv(os.path.join(path, 'Prepared Data', 'orders_checked1.csv'))

In [24]:
df_prods_clean.to_csv(os.path.join(path, 'Prepared Data', 'products_checked1.csv'))